src/range.cr

```diff
@@ -570,6 +570,101 @@ struct Range(B, E)
       end
     end
   end
+
+  # Clamps `self` within the boundaries of `range`.
+  #
+  # The result is the intersection of both ranges.
+  #
+  # ```
+  # (-5..5).clamp(0..100) # => 0..5
+  # (-5..5).clamp(-10..2) # => -5..2
+  # (-5..5).clamp(0..nil) # => 0..5
+  # ```
+  #
+  # If either range is exclusive, the result is exclusive, too.
+  # If both ranges are disparate, or have different `#direction` the result is
+  # an empty range expressed as `begin...begin` (or `end...end` if `begin` is
+  # `nil`).
+  #
+  # ```
+  # (-5...5).clamp(0..100) # => 0...5
+  # (-5..5).clamp(0...100) # => 0...5
+  # (-5..5).clamp(10..20)  # => -5...-5
+  # (-5..5).clamp(5..-5)   # => -5...-5
+  # (nil..0).clamp(1..nil) # => 0...0
+  # ```
+  def clamp(range : Range) : Range
+    range_direction = range.direction
+    direction = self.direction
+
+    if direction != range_direction && !direction.zero? && !range_direction.zero?
+      # result is an empty range and gets normalized
+      v = (self.begin || self.end).not_nil!
+      return v...v
+    end
+
+    if range_direction < 0
+      # if both ranges are descending, reverse range in order to work for `Number#clamp` calls
+      range = range.reverse
+    end
+
+    clamped_begin = self.begin.try &.clamp(range.begin, range.end) || range.begin
+    clamped_end = self.end.try &.clamp(range.begin, range.end) || range.end
+
+    exclusive = exclusive? || range.exclusive?
+
+    if (clamped_begin == clamped_end) && (clamped_begin != self.begin) && (clamped_end != self.end) && !range_direction.zero?
+      # result is an empty range and gets normalized
+      v = (self.begin || self.end).not_nil!
+      return v...v
+    end
+
+    case {clamped_begin, clamped_end}
```

Comment:
I'm probably missing something here. Is there a reason we need to use `case` rather than just always doing this?
```crystal
Range.new(clamped_begin, clamped_end, exclusive)
```

Reply:
It seems this is necessary to avoid nilable generic arguments. For example, without this the return type of `typeof((0..nil).clamp(nil..5))` is `(Range(Int32 | Nil, Int32 | Nil) | Range(Int32, Int32))`. I'm not 100% confident that this is the best solution but it's the best I could come up with for now.